### DISCUSSION TITLE
Is test suites running in parallel by default?

### DISCUSSION THREAD

**Discussion by xsjcTony** [vitest-dev/vitest#2668] [Q&A]
Hey, I just wanted to know if suites (`describe()`) are running in parallel by default (**including nested suites**) or do we have some settings?

I didn't find any corresponding docs related to that, if I use `describe.concurrent()`, it's going to mark all tests inside that suite as concurrent, but seems it does nothing to the suite it self, so I came up with this question.

---

And also, am I correct about `describe.concurrent` will only mark it's own test cases as concurrent, while it has no effect to nested suite inside?

---

Thanks for the reply!

**Comment by gpoole** [comment]
An old question and I think the docs have been improved since this was asked, but I'll leave the answer here for anyone coming from search like me. [Current documentation for `describe.concurrent`](https://vitest.dev/api/#describe-concurrent) makes clear and shows an example: using `describe.concurrent` makes _all_ tests under that suite at any level concurrent with each other.

One caveat is that using `describe.concurrent` on two suites at the top level doesn't make the _suites themselves_ run concurrently with each other. For example, here "suite a" and "suite b" will always run sequentially, even though their inner tests run concurrently:

```ts
// Runs first
describe.concurrent('suite a', () => {
  it("test a", async () => {
    // runs concurrently with test b
  });
  it("test b, async () => {
    // runs concurrently with test a
  });
});

// Runs second
describe.concurrent('suite b', () => {
  it("test c", async () => {
    // runs concurrently with test d
  });
  it("test d", async () => {
    // runs concurrently with test c
  });
});
```

One solution is to use an outer concurrent suite, for example:

```ts
describe.concurrent("suite", () => {
  describe('suite a', () => {
    it("test a", async () => {
      // runs concurrently with every other test
    });
    it("test b, async () => {
      // runs concurrently with every other test
    });
  });
  
  describe('suite b', () => {
    it("test c", async () => {
      // runs concurrently with every other test
    });
    it("test d", async () => {
      // runs concurrently with every other test
    });
  });
});
```

You can also make all tests in a file be concurrent with [the `sequence.concurrent` config option](https://vitest.dev/config/#sequence-concurrent) or [`--sequence-concurrent`](https://vitest.dev/guide/cli.html#sequence-concurrent) CLI flag.

You can opt back in to sequential execution for a nested suite by using [`describe.sequential`](https://vitest.dev/api/#describe-sequential). Unfortunately the docs don't clarify whether `sequential` also goes to any depth, but seems reasonable to assume.
